controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -7,51 +7,50 @@
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
  */
-package io.pravega.controller.server.retention;
+package io.pravega.controller.server.bucket;
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.util.concurrent.AbstractService;
-import io.netty.util.internal.ConcurrentSet;
 import io.pravega.common.concurrent.Futures;
-import io.pravega.common.tracing.RequestTracker;
-import io.pravega.controller.store.stream.StreamMetadataStore;
-import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.store.stream.BucketStore;
 import io.pravega.controller.util.RetryHelper;
 import lombok.extern.slf4j.Slf4j;
 
 import java.util.Collections;
-import java.util.Set;
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 @Slf4j
-public class StreamCutService extends AbstractService implements BucketOwnershipListener {
-    private final int bucketCount;
+public class BucketManager extends AbstractService implements BucketOwnershipListener {
     private final String processId;
-    private final ConcurrentSet<StreamCutBucketService> buckets;
-    private final StreamMetadataStore streamMetadataStore;
-    private final StreamMetadataTasks streamMetadataTasks;
+    private final BucketStore.ServiceType serviceType;
+    private final BucketStore bucketStore;
+    private final Function<Integer, AbstractBucketService> bucketServiceSupplier;
+    private final ConcurrentMap<Integer, AbstractBucketService> buckets;
     private final ScheduledExecutorService executor;
-    private final RequestTracker requestTracker;
+    private final Object bucketOwnershipLock = new Object();
 
-    public StreamCutService(final int bucketCount, String processId, final StreamMetadataStore streamMetadataStore,
-                            final StreamMetadataTasks streamMetadataTasks, final ScheduledExecutorService executor,
-                            final RequestTracker requestTracker) {
-        this.bucketCount = bucketCount;
+    BucketManager(final String processId, final BucketStore bucketStore,
```

Comment:
please add `@NonNull` to each argument to ensure that nothing that shouldn't be `null` isn't.